apps/www/_blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization.mdx

```diff
@@ -0,0 +1,208 @@
+
+To achieve Realtime authorization, we looked into our current solutions, namely how [Storage](https://supabase.com/docs/guides/storage/security/access-control) handles Access Control. Due to the nature of Realtime, our primitives are different as we have no assets stored in the database. So how did we achieve it?
+
+On Channel subscription you are able to inform Realtime to use a private Channel and we will do the required checks.
+
+The checks are done by running SELECT and INSERT queries on the new `realtime.messages` table which are then rolled backed so nothing is persisted. Then, based on the query result, we can determine the policies the user has for a given extension.
+
+As a result, in the server, we create a map of policies per connected socket so we can keep them in memory associated with the user's connection.
+
+```elixir
+%Policies{
+  broadcast: %BroadcastPolicies{read: false, write: false},
+  presence: %PresencePolicies{read: false, write: false}
+}
+```
+
+### One user, one context, one connection
+
+Now that we have set up everything on the database side, let's understand how it works and how we can verify authorization via RLS policies.
+
+Upon channel subscription, based on the configuration message sent when joining the websocket, we determine that you want to use a private channel, then we parse the information from your signed JSON Web Token (JWT), load them into a Postgres transaction using `set_config`, and verify them against the new `realtime.messages`table within the same transaction and we store the output as a group of policies within the context of the user's channel process.
+
+<Img
+  src={{
+    dark: '/images/blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--dark.png',
+    light:
+      '/images/blog/2024-08-13-supabase-realtime-broadcast-and-presence-authorization/one-user-connection--light.png',
+  }}
+  alt="Flow of checks done against the databse to determine the policies for a user connection."
+/>
+## How does this scale?
+
+Since we're doing this as soon as the user subscribes to Channel we should see a small latency increase in subscriptions but after that we do not run any queries on the database and we use the stored context to control the actions a user can perform.
+
+Latency between geographically close users is very important for a product like Realtime. To deliver messages as fast as possible between users on our global network, we cache the policies.
+
+We can maintain high throughput and low latency on a Realtime Channel with Broadcast and Presence authorization because:
+
+- the policy is only generated when a user connects to a Channel
+- the policy cached in memory is close to your users
+- the policy is cached for the duration of the connection or until the JWT expires
+
+Another added bonus is the fact that if a user does not have access to a given Channel they won't be able to connect at all and their connections will be rejected.
```

Comment:
I would take this out. It's not an added bonus; it should be core to how it works.

```suggestion
```

Reply:
I will change it to `If a user does not have access to a given Channel they won't be able to connect at all and their connections will be rejected.`

I do think it's relevant to keep something here regarding the fact that failed policies result in non connection and, by consequence, no connected user (which is part of our billing and could be a concern)